crates/sui-core/src/authority_aggregator.rs

```diff
@@ -633,29 +644,99 @@ where
         authority_errors: &mut HashMap<AuthorityName, SuiError>,
     ) -> Result<S, SuiError>
     where
-        FMap: Fn(AuthorityName, &'a SafeClient<A>) -> AsyncResult<'a, S, SuiError>,
+        FMap: Fn(AuthorityName, SafeClient<A>) -> AsyncResult<'a, S, SuiError> + Send + Clone + 'a,
         S: Send,
     {
+        let start = tokio::time::Instant::now();
         let mut delay = Duration::from_secs(1);
         loop {
             let authorities_shuffled = self.committee.shuffle_by_stake(preferences, restrict_to);
+            let mut authorities_shuffled = authorities_shuffled.iter();
+
+            type RequestResult<S> = Result<Result<S, SuiError>, tokio::time::error::Elapsed>;
+
+            enum Event<S> {
+                StartNext,
+                Request(AuthorityName, RequestResult<S>),
+            }
 
-            // TODO: possibly increase concurrency after first failure to reduce latency.
-            for name in authorities_shuffled {
-                let client = &self.authority_clients[&name];
+            let mut futures = FuturesUnordered::<BoxFuture<'a, Event<S>>>::new();
 
-                let res = timeout(timeout_each_authority, map_each_authority(name, client)).await;
+            let start_req = |name: AuthorityName, client: SafeClient<A>| {
+                let map_each_authority = map_each_authority.clone();
+                Box::pin(async move {
+                    trace!(?name, now = ?tokio::time::Instant::now() - start, "new request");
+                    let map = map_each_authority(name, client);
+                    Event::Request(name, timeout(timeout_each_authority, map).await)
+                })
+            };
 
+            let schedule_next = || {
+                let delay = self.timeouts.serial_authority_request_interval;
+                Box::pin(async move {
+                    sleep(delay).await;
+                    Event::StartNext
+                })
+            };
+
+            // This process is intended to minimize latency in the face of unreliable authorities,
+            // without creating undue load on authorities.
+            //
+            // The fastest possible process from the
+            // client's point of view would simply be to issue a concurrent request to every
+            // authority and then take the winner - this would create unnecessary load on
+            // authorities.
+            //
+            // The most efficient process from the network's point of view is to do one request at
+            // a time, however if the first validator that the client contacts is unavailable or
+            // slow, the client must wait for the serial_authority_request_timeout period to elapse
+            // before starting its next request.
+            //
+            // So, this process is designed as a compromise between these two extremes.
+            // - We start one request, and schedule another request to begin after
```

Comment:
It is still inefficient to timeout on every request, even if the timeout to contact a second authority is shorter than the full timeout. Have a look at the `AuthorityHealth` information that is stored within the `ActiveAuthority` and shared between all active services of a node. It allows the active services upon failure / timeout or byz behaviour to register the event as a minimum amount of time before contacting this authority again. This way upon the first failure to respond we can remember to not contact an authority again for some time, or prefer others.

Feel free to augment the Health information to suit your needs too.

Reply:
Wow, wasn't aware of that.

The AuthorityHealth should really be used by every part of the AuthorityAggregator, but that's more than enough work for another PR. I will create an issue so we don't lose track of this item.

Reply:
Filed #3515